h5py/_hl/vds.py

```diff
@@ -156,17 +160,83 @@ class VirtualLayout(object):
     maxshape
         The virtual dataset is resizable up to this shape. Use None for
         axes you want to be unlimited.
+    filename
+        The name of the destination file, if known in advance. Mappings from
+        data in the same file will be stored with filename '.', allowing the
+        file to be renamed later.
     """
-    def __init__(self, shape, dtype, maxshape=None):
+    def __init__(self, shape, dtype, maxshape=None, filename=None):
         self.shape = (shape,) if isinstance(shape, int) else shape
         self.dtype = dtype
         self.maxshape = (maxshape,) if isinstance(maxshape, int) else maxshape
-        self.sources = []
+        self.filename = filename
+        self._src_filenames = set()
+        self.dcpl = h5p.create(h5p.DATASET_CREATE)
 
     def __setitem__(self, key, source):
         sel = select(self.shape, key, dataset=None)
         _convert_space_for_key(sel.id, key)
-        self.sources.append(VDSmap(sel.id,
-                                   source.path,
-                                   source.name,
-                                   source.sel.id))
+        src_filename = self._source_file_name(source.path, self.filename)
+
+        self.dcpl.set_virtual(
+            sel.id, src_filename, source.name.encode('utf-8'), source.sel.id
+        )
+        if self.filename is None:
+            self._src_filenames.add(src_filename)
+
+    @staticmethod
+    def _source_file_name(src_filename, dst_filename) -> bytes:
+        src_filename = filename_encode(src_filename)
+        if dst_filename and (src_filename == filename_encode(dst_filename)):
+            # use relative path if the source dataset is in the same
+            # file, in order to keep the virtual dataset valid in case
+            # the file is renamed.
+            return b'.'
+        return filename_encode(src_filename)
+
+    def get_dcpl(self, dst_filename=None):
+        """Get the property list containing virtual dataset mappings
+
+        If the destination filename wasn't known when the VirtualLayout was
+        created, it is handled here.
+        """
+        dst_filename = filename_encode(dst_filename)
+        if (self.filename is None) and (dst_filename in self._src_filenames):
```

Comment:
I'm not sure that the complexity of _sometimes_ knowing the destination file up front is worth it?

Reply:
I'm also worried about bug reports where someone changes `layout.filename` after having created it and now has very mangled VDS if we thought we could use `b'.'` when `__setitem__` was called, but not when `make_dataset` was called.

Also, if I'm reading this right I think that if `parent.file.filename != layout.filename` we can also get mangled VDS?

Reply:
So part of the motivation is #1865, where @ajelenak complained about the memory usage of a large number of mappings, because both h5py & HDF5 store them. I think that's a somewhat pathological case, but it's not much extra complexity to handle when we do know the filename in advance. My use cases could plausibly have a few thousand mappings, and I know calls into HDF5 have some overhead, so I'd rather not copy everything across to a new plist if I don't have to.

I agree that there's scope for messing it up. How about if I make `_filename` and `_get_dcpl` private, and add a check that `dst_filename == self._filename` (unless `self._filename is None`)? The intention is that if you give it a filename up front, you've committed to that as the destination file.

Reply:
I guess we might also make the filename matching and translation to `'.'` optional - I'm not sure if that helps or just adds another dimension of complication.

Reply:
I had not considered the hdf5 call overhead.

Should this check be a bit more complicated? I think we need to go through making a new plist in this case (where the layout did not pre-know its file name and the dest is in the sources) and in the case where it did know the file name, but it was different than `dst_filename` ultimately was.  If the name-the-layout-know is in `self._src_filenames` then we need to both convert some `b'.'` -> paths and if dst_filename is in `self.src_filenames` then we may get to convert some paths ot `b'.'` 

We probably should make `self.filename` private as well?

Reply:
> I agree that there's scope for messing it up. How about if I make _filename and _get_dcpl private, and add a check that dst_filename == self._filename (unless self._filename is None)? The intention is that if you give it a filename up front, you've committed to that as the destination file.

I like that plan.

> I guess we might also make the filename matching and translation to '.' optional - I'm not sure if that helps or just adds another dimension of complication.

Maybe wait until someone complain about not wanting relative paths?

Reply:
[deleted a comment which I wrote having not read @takluyver 's second paragraph 🤦🏻 ]